### ISSUE TITLE
hidapi is not working properly on FreeBSD

### ISSUE THREAD

**Issue by HenryHu** [libsdl-org/SDL#5700]
I tried to use SDL2 with a Google Stadia controller and noticed that it's not recognized by the 'better' hidapi drivers. It's recognized as a uhid device by the native driver - but that driver only recognizes the exported gamepad interface, and does not support some functionalities, such as the analog triggers.

I traced this to improper support of libusb on FreeBSD. It seems like that on FreeBSD, SDL_LIBUSB_DYNAMIC is false, because HIDAPI_ONLY_LIBUSB is true. However, that results in SDL_LIBUSB_DYNAMIC not defined, which is necessary for src/hidapi/SDL_hidapi to work properly.

I have some local changes which gets rid of the dynamic loading aspect of libusb - and the changes are surprisingly small - and it works. Now it can be detected (using the jstest-sdl info from github.com/meleu/jstest-sdl):

Found 4 joystick(s)

Joystick Name:     'Google Stadia Controller'
Joystick GUID:     03000000d11800000094000000016800
Joystick Number:    0
Number of Axes:     6
Number of Buttons: 17
Number of Hats:     0
Number of Balls:    0
GameController:
  Name:    'Google Stadia Controller'
  Mapping: '03000000d11800000094000000016800,*,a:b0,b:b1,back:b4,dpdown:b12,dpleft:b13,dpright:b14,dpup:b11,guide:b5,leftshoulder:b9,leftstick:b7,lefttrigger:a4,leftx:a0,lefty:a1,rightshoulder:b10,rightstick:b8,righttrigger:a5,rightx:a2,righty:a3,start:b6,x:b2,y:b3,misc1:b15,platform:FreeBSD'

Joystick Name:     '/dev/uhid2'
Joystick GUID:     2f6465762f7568696432000000000000
Joystick Number:    1
Number of Axes:     3
Number of Buttons:  8
Number of Hats:     0
Number of Balls:    0
GameController:
  not a gamepad

Joystick Name:     '/dev/uhid5'
Joystick GUID:     2f6465762f7568696435000000000000
Joystick Number:    2
Number of Axes:     3
Number of Buttons: 16
Number of Hats:     0
Number of Balls:    0
GameController:
  not a gamepad

Joystick Name:     '/dev/uhid8'
Joystick GUID:     2f6465762f7568696438000000000000
Joystick Number:    3
Number of Axes:     4
Number of Buttons: 15
Number of Hats:     1
Number of Balls:    0
GameController:
  not a gamepad

Notice that the last one (/dev/uhid8) is the native device, which only has 4 axies. The hidapi-based one has 6.
I'll try to see what's the best way to fix this. Maybe we can enable dynamic binding on FreeBSD as well.

**Comment by HenryHu** [comment]
I think I have a fix (https://github.com/libsdl-org/SDL/compare/main...HenryHu:main), however I'm not sure what's the best way in defining HAVE_LIBUSB in configure. I can change configure.ac, but the regenerated configure is too different from the one in the repo. What's the recommended way?

**Comment by slouken** [comment]
Go ahead and submit a pull request without configure regenerated, and we'll update that here.

Thanks!

**Comment by edenist** [comment]
Chiming in here to shed some more light on this issue. I can confirm that this appears to affect multiple [all?] hidapi devices on FreeBSD, as I have been trying to solve the exact same problem with PS4 controllers for some time now. 

@HenryHu I have applied your patches but this doesn't appear to make any difference with PS4 controllers. Are you sure the commits in your repo reflect all of the changes to fix things for you? My output from jstest is still:

```
Found 1 joystick(s)

Joystick Name:     '/dev/uhid0'
Joystick GUID:     2f6465762f7568696430000000000000
Joystick Number:    0
Number of Axes:     6
Number of Buttons: 14
Number of Hats:     1
Number of Balls:    0
GameControllerConfig:
  missing (see 'gamecontrollerdb.txt' or SDL_GAMECONTROLLERCONFIG)
```

If it helps, this is a regression which occurred sometime after 2.0.10 [2.0.12 failed due an error with missing hidapi.c, and even after this was solved in 2.0.18 I have been unable to get PS4 controllers working since]. For comparison, here is the output of jstest with 2.0.10:

```
Found 2 joystick(s)

Joystick Name:     '/dev/uhid0'
Joystick GUID:     2f6465762f7568696430000000000000
Joystick Number:    0
Number of Axes:     6
Number of Buttons: 14
Number of Hats:     1
Number of Balls:    0
GameControllerConfig:
  missing (see 'gamecontrollerdb.txt' or SDL_GAMECONTROLLERCONFIG)

Joystick Name:     'PS4 Controller'
Joystick GUID:     030000004c050000c405000000016800
Joystick Number:    1
Number of Axes:     6
Number of Buttons: 16
Number of Hats:     0
Number of Balls:    0
GameControllerConfig:
  Name:    'PS4 Controller'
  Mapping: '030000004c050000c405000000016800,*,a:b0,b:b1,back:b4,dpdown:b12,dpleft:b13,dpright:b14,dpup:b11,guide:b5,leftshoulder:b9,leftstick:b7,lefttrigger:a4,leftx:a0,lefty:a1,rightshoulder:b10,rightstick:b8,righttrigger:a5,rightx:a2,righty:a3,start:b6,x:b2,y:b3,'
```

I'll keep investigating  this issue as I'm now putting some time into getting it solved [I don't know how much longer I can keep kicking 2.0.10 along on my system without breaking something else, heh]. Thanks for your effort so far @HenryHu, it's been a big help on knowing where to look.

**Comment by edenist** [comment]
@HenryHu My apologies as it seemed my original comments were incorrect. I have now successfully built SDL with your patches and confirmed this also fixes hidapi support for PS4 as well. jstest output now mirrors that of 2.0.10! 

What I did wrong though is still somewhat relevant... I was building using configure instead cmake. It wasn't until I realised your changes included modifications to the cmake files that I had my head-slap moment and rebuilt with cmake. The build scripts for  FreeBSD ports currently use configure, so that was what I was used to. Presumably as part of the 2.0.26 milestone this change will be worked on to give parity between configure and cmake builds. 

Nevertheless, the main points I wanted to make are: 
1) Confirming these changes fix PS4 controller support, thus likely all hidapi controllers too
2) currently cmake only
3) Thanks immensely @HenryHu! I've been running 2.0.10 since 2019 for this very reason!
